### DISCUSSION TITLE
Ignore/preserve version classifier on update in specific Maven dependencies

### DISCUSSION THREAD

**Discussion by szpak** [renovatebot/renovate#19747] [Request Help] [answered]
Spock framework is a testing framework for Java and Groovy. It has [versions](https://search.maven.org/artifact/org.spockframework/spock-core) compiled for different Groovy versions (for binary/runtime compatibiilty). For example:

```
//Spock 2.2 - artifact: spock, version: 2.2, version-classifier: `-groovy-x.y`
spock-2.2-groovy-2.5
spock-2.2-groovy-3.0
spock-2.2-groovy-4.0

//Spock 2.3
spock-2.3-groovy-2.5
spock-2.3-groovy-3.0
spock-2.3-groovy-4.0
```

Usually, it doesn't make sense to propose update from `spock-2.2-groovy-3.0` to `spock-2.2-groovy-4.0` as (unless, the Groovy version used in the project is also updated - however, it is quick tricky to detect, so I would ignore that sophisticated mechanism). It would be useful to be able to define a rule to bump ignore (to preserve?) the **version** suffix (here `-groovy-x.y`) and for example for `spock-2.2-groovy-3.0` propose and update to `spock-2.3-groovy-3.0` (but NOT to `spock-2.2-groovy-4.0` nor `spock-2.3-groovy-4.0`).

```
spock-2.2-groovy-3.0 → spock-2.3-groovy-3.0  //ok
spock-2.2-groovy-3.0 → spock-2.3-groovy-4.0  //no ok - compile/runtime errors
spock-2.3-groovy-3.0 → spock-2.3-groovy-4.0  //no ok - compile/runtime errors
```

Is it possible to ignore/preserve the **version** suffix for selected packages, using the `packageRules`?

**Comment by szpak** [comment]
Of course, I've found [this](https://github.com/renovatebot/renovate/discussions/17560#discussioncomment-3526779) right after sending my question...

**Comment by szpak** [comment]
Using:
```
  "packageRules": [
    {
      "matchPackagePrefixes": ["org.spockframework:spock-"],
      "allowedVersions": "/-groovy-3\\.0$/"
    }
  ]
```
works. However, due to the usage of Groovy 3 and 4 in different project, I cannot use it in the global config.

Maybe there is a way to just "keep" some suffix and treat the other part as a version?

For example:
> spock-2.2(-groovy-3.0) → spock-2.3(-groovy-3.0) → spock-2.4(-groovy-3.0)

and
> spock-2.2(-groovy-4.0) → spock-2.3(-groovy-4.0) → spock-2.4(-groovy-4.0)

**Reply by rarkins** [reply]
The use of a hyphen inside the suffix itself makes this hard to do. I think you'll probably need one rule per groovy version

**Reply by viceice** [reply]
maybe you can combine it with `matchCurrentValue`

**Comment by szpak** [comment, marked as answer]
Thanks @viceice for the suggestion! The following - placed in the global config - seems to work for both Groovy 3.0 and 4.0:

```json5
packageRules: [
    ...
    {
      matchPackagePrefixes: ["org.spockframework:spock-"],
      matchCurrentVersion: "/-groovy-3\\.0$/",
      allowedVersions: "/-groovy-3\\.0$/"
    },
    {
      matchPackagePrefixes: ["org.spockframework:spock-"],
      matchCurrentVersion: "/-groovy-4\\.0$/",
      allowedVersions: "/-groovy-4\\.0$/"
    },
]
```
